### ISSUE TITLE
Multithread-Capable MDC

### ISSUE THREAD

**Issue by darkfrog26** [outr/scribe#63]
MDC sucks as a paradigm on Logback and SLF4J as it's a single-threaded / ThreadLocal nightmare and entirely useless in asynchronous operations.  We should add a better core system (perhaps building upon `ExecutionContext`: #62 ) to easily and conveniently support thread-safe usage.

**Comment by kell18** [comment]
Hi! Thanks for the lib! As far as I can see it still will suffer from context switches? I.e. when a function executing in the future is getting moved to another execution thread? The MDC context for the function will not be preserved, right?

**Comment by darkfrog26** [comment]
So, this was a bit of a half-measure. In order to share the context, you'd have to `MDC.contextualize(mdc) { ... }` in the block you want to utilize it. I'll give this some more thought on coming up with a better way to do it. If you have any thoughts, they would be welcome.

**Comment by darkfrog26** [comment]
What would you think about me adding:
```scala
(implicit mdc: MDC = MDC.global)
```
to all logging calls and allow an implicit MDC to be able to be picked up?

**Comment by kell18** [comment]
Hmm, not many ideas from my side. But maybe it's not the task for the logging library if it was not initial design choice. I heard that it should be possible to do with custom ExecutionContexts but I never was able to make it working... https://blog.knoldus.com/how-to-use-mdc-logging-in-scala/

**Comment by darkfrog26** [comment]
I've added some new functionality to allow the use of implicits:
```scala
    MDC { implicit mdc =>
      mdc("key") = "value"
      val f = Future {
        scribe.info("this is a test")
      }
      Await.result(f, Duration.Inf)
    }
```
The above will output with "key" = "value" in the logs. Would you say this adequately solves the problem? Of course, if you're working across multiple methods you'd have to pass around the implicit MDC.

**Comment by darkfrog26** [comment]
@kell18 I'm planning to do a release this week, but I'd like confirmation that this update fulfills your needs?

**Comment by darkfrog26** [comment]
Got tired of waiting. Released with 3.10.0
